server/app/Icepeak/Main.hs

```diff
@@ -77,33 +74,54 @@ runCore core = do
   httpServer <- HttpServer.new core
   let wsServer = WebsocketServer.acceptConnection core
 
+  let tryWithTag wrapper act = catch (act >> pure RCR_Fine) $ \exc -> case fromException exc of
+        Just (_ :: AsyncException) -> pure $ RCR_Fine
+        _ -> pure $ wrapper exc
+
   -- start threads
-  pops <- Async.async $ Core.runCommandLoop core
+  pops <- Async.async $ tryWithTag RCR_CommandLoop $ Core.runCommandLoop core
+  upds <- Async.async $ tryWithTag RCR_WebSockets $ WebsocketServer.processUpdates core
+  serv <- Async.async $ tryWithTag RCR_HTTP $ Server.runServer logger wsServer httpServer
   sync <- Async.async $ Core.runSyncTimer core
-  upds <- Async.async $ WebsocketServer.processUpdates core
-  serv <- Async.async $ Server.runServer logger wsServer httpServer
   metrics <- Async.async
     $ forM_ (configMetricsEndpoint config) (MetricsServer.runMetricsServer logger)
-  installHandlers core serv
+
+  installHandlers core
   logAuthSettings config logger
   logQueueSettings config logger
   logSyncSettings config logger
   postLog logger "System online. ** robot sounds **"
 
-  waitLog "core" logger pops
-  waitLog "web sockets server" logger upds
-  waitLog "http server" logger serv
-  -- kill auxiliary threads when the main threads ended
+  -- Everything should stop when any of these stops
+  (_, runCoreResult) <- Async.waitAny [pops, upds, serv]
+  logRunCoreResult logger runCoreResult
+
+  -- kill all threads when one of the main threads ended
+  Core.postQuit core -- should stop pops
+  Async.cancel upds
+  Async.cancel serv
   Async.cancel metrics
   Async.cancel sync
-
--- | Wait for an Async computation to exit and log unexpected exceptions.
-waitLog :: Text -> Logger -> Async () -> IO ()
-waitLog name logger action = Async.waitCatch action >>= handleLog where
-  handleLog (Left exc)
-    | Just (_ :: AsyncException) <- fromException exc = pure ()
-    | otherwise = Logger.postLog logger $ name <> " stopped with an exception: " <> Text.pack (show exc)
-  handleLog (Right _) = pure ()
+  void $ Async.wait pops
```

Comment:
Maybe use `forM_` to kill all threads.

Reply:
Skipping this one because it doesn't really read better:
```
  mapM_ Async.cancel [ void webSocketThread
                     , void httpThread
                     , metricsThread
                     , syncThread
                     ]
```